### PR TITLE
Add CDF reader to TimeSeries

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/heliopython/heliopy/issues/949

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -61,6 +61,7 @@ net =
   tqdm>=4.32.1
   zeep>=3.4.0
 timeseries =
+  cdflib>=0.3.20
```
Comment (dstansby) [inline setup.cfg:64]: Need to check this is a suitable version to pin to, and add a corresponding entry to the oldesttestdeps CI.

File: sunpy/io/cdf.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import cdflib
```
Comment (dstansby) [inline sunpy/io/cdf.py:1]: Need to add this file to the docs.

File: sunpy/io/cdf.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                units[var_key] = unit
+
+        # TODO: extract some sort of metadata from the CDF file
+        all_ts.append(GenericTimeSeries(data=df, units=units, meta=meta))
+
+    return all_ts
+
+
+if __name__ == '__main__':
+    read_cdf('/Users/dstansby/Downloads/psp_fld_l2_mag_rtn_1min_20200104_v02.cdf')
```
Comment (dstansby) [inline sunpy/io/cdf.py:83]: Lol this needs to go.

File: sunpy/io/cdf.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                # Multiple columns, give each column a unique label
+                for i, col in enumerate(data.T):
+                    df[var_key + f'_{i}'] = col
+                    units[var_key + f'_{i}'] = unit
+            else:
+                # Single column
+                df[var_key] = data
+                units[var_key] = unit
+
+        # TODO: extract some sort of metadata from the CDF file
```
Comment (dstansby) [inline sunpy/io/cdf.py:76]: Remove me

File: sunpy/timeseries/timeseries_factory.py [hunk 5]
```diff
@@ -138,6 +138,15 @@ def _read_file(fname, **kwargs):
             List of ``(data, header)`` pairs or ``fname`` if the file is not supported or incorrect.
         """
         if 'source' not in kwargs.keys() or not kwargs['source']:
+            try:
+                if detect_filetype(fname) == 'cdf':
+                    # Put import here to ensure there is no import dependency
+                    # on cdflib for TimeSeries
+                    from sunpy.io.cdf import read_cdf
+                    return read_cdf(os.fspath(fname), **kwargs)
+            except UnrecognizedFileTypeError:
```
Comment (dstansby) [inline sunpy/timeseries/timeseries_factory.py:147]: Also need to except an importerror

**Review by dstansby — COMMENTED** [review]
TODO:
- Add a what's new entry

**Review by nabobalis — APPROVED** [review]

**Comment by Cadair** [comment]
This PR should have had two reviews as it's not a bugfix.

That being said it looks good :partying_face:

**Pull request merged**